drivers/ethernet/phy_xlnx_gem.c

```diff
@@ -0,0 +1,975 @@
+/*
+ * Xilinx Processor System Gigabit Ethernet controller (GEM) driver
+ *
+ * PHY management interface implementation
+ * Models currently supported:
+ * - Marvell Alaska 88E1111 (QEMU simulated PHY)
+ * - Marvell Alaska 88E1510/88E1518/88E1512/88E1514 (Zedboard)
+ * - Texas Instruments TLK105
+ * - Texas Instruments DP83822
+ *
+ * Copyright (c) 2021, Weidmueller Interface GmbH & Co. KG
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <device.h>
+
+#include "eth_xlnx_gem_priv.h"
+
+#define LOG_MODULE_NAME phy_xlnx_gem
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+/* Basic MDIO read / write functions for PHY access */
+
+/**
+ * @brief Read PHY data via the MDIO interface
+ * Reads data from a PHY attached to the respective GEM's MDIO interface
+ *
+ * @param base_addr Base address of the GEM's register space
+ * @param phy_addr  MDIO address of the PHY to be accessed
+ * @param reg_addr  Index of the PHY register to be read
+ * @return          16-bit data word received from the PHY
+ */
+static uint16_t phy_xlnx_gem_mdio_read(
+	uint32_t base_addr, uint8_t phy_addr,
+	uint8_t reg_addr)
+{
+	uint32_t reg_val = 0;
+	uint32_t poll_cnt = 0;
+
+	/*
+	 * MDIO read operation as described in Zynq-7000 TRM,
+	 * chapter 16.3.4, p. 517.
+	 */
+
+	/*
+	 * Wait until gem.net_status[phy_mgmt_idle] == 1 before issuing the
+	 * current command.
+	 */
+	do {
+		if (poll_cnt++ > 0)
+			k_busy_wait(100);
+		reg_val = sys_read32(base_addr + ETH_XLNX_GEM_NWSR_OFFSET);
+	} while ((reg_val & ETH_XLNX_GEM_MDIO_IDLE_BIT) == 0 && poll_cnt < 10);
+	if (poll_cnt == 10) {
+		LOG_ERR("GEM@0x%08X read from PHY address %hhu, "
+			"register address %hhu timed out",
+			base_addr, phy_addr, reg_addr);
+		return 0;
+	}
+	poll_cnt = 0;
```

Comment:
i'd move this down to the do-while below, just in case code gets added between here and there in the future.